# **I Solving the Sum of Two Digits Programming Challenge**

## C++

In [2]:
%%writefile AplusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing AplusB.cpp


In [3]:
!g++ AplusB.cpp -o AplusB
!./AplusB

4 4
8

### Test input output files

In [4]:
%%writefile input1.txt
4 10

Writing input1.txt


In [5]:
%%writefile output1.txt
14

Writing output1.txt


In [6]:
%%writefile input2.txt
-5 4

Writing input2.txt


In [7]:
%%writefile output2.txt
-1

Writing output2.txt


In [8]:
!./AplusB < input1.txt

14

In [9]:
!./AplusB < input2.txt

-1

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [10]:
!grep -xvFf output1.txt output1.txt 

In [11]:
!grep -xvFf output1.txt output2.txt 

-1


In [12]:
!grep -xvFf output2.txt output1.txt 

14


In [13]:
!./AplusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [14]:
!./AplusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [15]:
%%writefile AplusB.py
# python3


def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing AplusB.py


In [16]:
!python AplusB.py

4 4
8


In [17]:
!python AplusB.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [18]:
!python AplusB.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# **II Solving The Maximum Pairwise Product Programming Challenge in C++**

### **First Solution**

#### **C++**

In [19]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product.cpp


In [20]:
!g++ max_pairwise_product.cpp -o max_pairwise_product
!./max_pairwise_product

10
1 2 3 4 5 6 7 8 9 10
90


##### Test 1: Small numbers

In [21]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [22]:
!./max_pairwise_product < input1.txt

90


##### Test 2: largest input numbers 200000 200000 in the input

In [23]:
%%writefile input2.txt
6
1 3 200000 9 200000 23

Overwriting input2.txt


In [24]:
!./max_pairwise_product < input2.txt

1345294336


Changing the variables to long long

In [25]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
			       aux_product =  (long long) numbers[first] * numbers[second];
            if ( max_product < aux_product)  {
				        max_product = aux_product;
             }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long.cpp


In [26]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!./max_pairwise_product_long < input1.txt
!./max_pairwise_product_long < input2.txt

90
40000000000


##### Test 3 stress tests 20 times random n in [2, 20]

In [27]:
%%writefile test3.cpp
#include <iostream>
#include <ctime>
using namespace std;
int main()
{
   srand(time(NULL));
   int i, n, aleatorio, DESDE=0, HASTA=200000;
   n = 20;
   std::cout<<n<<"\n";
   for (i = 1; i <= n; i ++)
   {
             aleatorio = rand()%(HASTA-DESDE+1)+DESDE;
             std::cout << aleatorio << " ";
   }
   std::cout<<"\n";
   system("pause");
}

Writing test3.cpp


In [28]:
!g++ test3.cpp -o test3
!./test3 > input3.txt

sh: 1: pause: not found


In [29]:
!time ./max_pairwise_product_long <input3.txt

35802189619

real	0m0.008s
user	0m0.002s
sys	0m0.002s


##### Test 4 time test measure the execution time running with the the largest worst case file (200000 increasing numbers, 200000 and 200000 the end)

In [30]:
%%writefile test4.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main() {
    int n=200001;
    std::cout<<n<<"\n";
    for (int i = 1; i <= n; ++i) {
        std::cout << i << " ";
    }
    return 0;
}

Writing test4.cpp


In [31]:
!g++ test4.cpp -o test4
!./test4 > input4.txt

Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [32]:
!time ./max_pairwise_product_long <input4.txt

40000200000

real	1m58.773s
user	1m57.704s
sys	0m0.054s


#### **Python**

Se va a realizar el mismo ejercicio utilizando Python

In [76]:
%%writefile max_pairwise_product.py
# python3

def max_pairwise_product(numbers):
    n = len(numbers)
    max_product = 0
    for first in range(n):
        for second in range(first + 1, n):
            max_product = max(max_product,
                numbers[first] * numbers[second])

    return max_product


if __name__ == '__main__':
    input_n = int(input())
    input_numbers = [int(x) for x in input().split()]
    print(max_pairwise_product(input_numbers))


Overwriting max_pairwise_product.py


##### Test 1: Small numbers

In [77]:
!time python max_pairwise_product.py<input1.txt

90

real	0m0.065s
user	0m0.044s
sys	0m0.015s


##### Test 2: largest input numbers 200000 200000 in the input

In [78]:
!time python max_pairwise_product.py<input2.txt

40000000000

real	0m0.068s
user	0m0.051s
sys	0m0.014s


##### Test 3 stress tests 20 times random n in [2, 20]

In [79]:
!time python max_pairwise_product.py<input3.txt

35802189619

real	0m0.068s
user	0m0.049s
sys	0m0.016s


##### Test 4 time test measure the execution time running with the the largest worst case file (200000 increasing numbers, 200000 and 200000 the end)

In [80]:
!time python max_pairwise_product.py<input4.txt

Traceback (most recent call last):
  File "max_pairwise_product.py", line 17, in <module>
    print(max_pairwise_product(input_numbers))
  File "max_pairwise_product.py", line 7, in max_pairwise_product
    for second in range(first + 1, n):
KeyboardInterrupt

real	9m15.199s
user	9m10.208s
sys	0m0.212s


**Number of instruction**

Lo que se va a hacer a continuación es analizar la cantidad de instrucciones que realiza el algoritmo para las 200000 entradas, viendo el código se puede ver que hay dos for, el for externo itera n veces, mientras que el for interno itera n-1 veces. Haciendo cuentas:

$$
n(n-1)=200.000(200.000-1)
$$
$$
= 39.999.800.000  \; \text{ instrucciones}
$$

**Time x Instruction**

Una vez con un aproximado de las operaciones realizadas, se calcula el tiempo que tarda haciendo cada instrucción, que es igual a:
$$
1h24m44s = 5.084s
$$
Por lo que:
$$
\frac{39.999.800.000 \;\; \text{instrucciones}}{5.084\;\;\;\text{segundos}}
$$
Que es igual a:
$$
7.867.781 \frac{\text{instrucciones}}{\text{segundo}}
$$

Lo que muestra una gran diferencia en rendimientos entre Python y C++

Se puede apreciar como el tiempo sobrepasa los límites, por lo que es necesario hacer un código más eficiente.

## **Max pairwise product execints**

##### Calculate  the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line 

```cpp
 int main() {                         //N inst exec by line executed  
   
    int n;                            //1                      
    std::cin >> n;                    //1
                                              
    std::vector<int> numbers(n);      //n    init vector
    .
    .
 }

```

In [ ]:
%%writefile test_init_vectors.cpp
#include <iostream>
#include <vector>
#include <algorithm>
int main() {                                                               
    int n = 10;                                                                                                                            //1
    std::vector<int> numbers(n);                                            
    for (int i = 0; i < n; ++i) {                                           
        std::cout << numbers[i];                                             
    }

    std::cout << "\n";                       
    return 0;                                                              
}  

Writing test_init_vectors.cpp


In [ ]:
!g++ test_init_vectors.cpp -o test_init_vectors
!./test_init_vectors

0000000000


In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed 
    long long max_product = 0LL;                                             //1                                          
	  long long aux_product = 0LL;                                             //1 
    int n = numbers.size();                                                  //1 
                                                                             
    for (int first = 0; first < n; ++first) {                                //1+2n  
        for (int second = first + 1; second < n; ++second) {                 //n + (n-1)n 
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2 
            if (max_product < aux_product)  {                                //(n-1)n/2 
				        max_product = aux_product;                                   //[0, (n-1)n/2] 
             }
        }
    }
                                                                             // -------------- 
    return max_product;                                                      // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] 
}

int main() {                                              //N inst exec by line executed 
    int n;                                                //1  
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
    return 0;                                             // -------------- 
}                                                         // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]

Overwriting max_pairwise_product_long.cpp


Count executed intructions in code and verify with inputs 1, 2 and 3

In [ ]:
%%writefile max_pairwise_product_long_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {              //N inst exec by line executed 
    
    long long m = 0LL;                                                       //1
    
    long long max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int n = numbers.size();                                                  //1 
    execints++;

    execints++;                                                                         
    for (int first = 0; first < n; ++first) {                                //1+2n  
        execints += 2;

        long long m = 0LL;

        execints++;                                    
        for (int second = first + 1; second < n; ++second) {                 //n+ (n-1)n 
            execints += 2;                                                

			      aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2 
            execints++; 

            execints++; 
            if ( max_product < aux_product)  {                               //(n-1)n/2 
                       
				        max_product = aux_product;                                   //[0, (n-1)n/2] 
                execints++; 
                m++;
             }
        }
        if ( (first <= 10) || (n- first <= 10)) std::cout << m << "\n";
    }
                                                                            // -------------- 
    return max_product;                                                     // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] 
}

int main() {                                                 //N inst exec by line executed 
    long long execints =0LL;

    int n;                                                   //1
    execints++;                                                              
    
    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;
    
    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n
    }
    
    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";                       
    return 0;                                                // -------------- 
}                                                            // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]

Writing max_pairwise_product_long_execints.cpp


In [ ]:
!g++ max_pairwise_product_long_execints.cpp -o max_pairwise_product_long_execints
!./max_pairwise_product_long_execints < input1.txt
!./max_pairwise_product_long_execints < input2.txt
!./max_pairwise_product_long_execints < input4.txt

9
5
4
3
2
2
2
2
1
0
289
2
1
2
0
0
0
116
200000
100001
66667
50001
40001
33334
28572
25001
22223
20001
18182
2
2
2
2
2
2
2
2
1
0
80004472127


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T = 9 + 7n + [2(n-1)n, 2.5 (n-1)n]$$

It depends on the size of the lists and  the order of the list.

For the input1.txt 
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 9 + 7(10) + 2(9)(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 289 $$

For the input2.txt 
```
6
1 3 200000 9 200000 23
```
$$T = 9 + 7(6) + 2(5)(6)  +  2 + 1 + 2 + 0 + 0 + 0 = 116 $$


For the input3.txt that is the worst case (n=200000 and list in incresing order)

```
200000
1 2 3 4 5 6 7 8 9 10 ... 199998 199999 200000
```

$$
\begin{array}{ll}
T &= 9 + 7(200000) + (2)(199999)(200000)  + (199999 +
100000 + 66667 +50000 \cdots 2 + 2 + 1)\\
&= 80003672077  \\
  &< 9 + 7(200000) + (2.5)(199999)(200000) \\
  & = 100000900009
\end{array}
  $$

Calculate estimated time by instruction and number of instructions by second

In [ ]:
# Estimated time by instruction

128/80003672077 


1.5999265618308827e-09

In [ ]:
# Estimated number of instructions by second

1/(128/80003672077) 

625028688.1015625

## **Efficient solution**

#### **C++**

Se plantean distintas soluciones eficientes al problema

In [43]:
%%writefile max_pairwise_product_eficiente.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <list>

long long MaxPairwiseProductEficiente(const std::vector<int>& numbers) {
    int n = numbers.size();
    long long max_product = 0LL;
    long long mayor = 0;
    long long smayor = 0;
    for(int i = 0; i < n; i++){
        if(numbers[i]>mayor){
          smayor = mayor;
          mayor = numbers[i];
        }else if(numbers[i]>smayor){
          smayor = numbers[i];
        }
    }
    max_product = (long long)mayor*smayor;
    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    long long maxproduc = 0LL;
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }
    maxproduc = MaxPairwiseProductEficiente(numbers);
    std::cout<<maxproduc<<"\n";
    return 0;
}



Writing max_pairwise_product_eficiente.cpp


In [44]:
!g++ max_pairwise_product_eficiente.cpp -o max_pairwise_product_eficiente

##### Test 1: Small numbers

In [48]:
!./max_pairwise_product_eficiente < input1.txt

90


##### Test 2: largest input numbers 200000 200000 in the input

In [49]:
!./max_pairwise_product_eficiente < input2.txt

40000000000


##### Test 3 stress tests 20 times random n in [2, 20]

In [54]:
!time ./max_pairwise_product_eficiente < input3.txt

35802189619

real	0m0.004s
user	0m0.002s
sys	0m0.002s


##### Test 4 time test measure the execution time running with the the largest worst case file (200000 increasing numbers, 200000 and 200000 the end)

In [53]:
! time ./max_pairwise_product_eficiente < input4.txt

40000200000

real	0m0.089s
user	0m0.077s
sys	0m0.005s


Lo que comprueba que es una solución eficiente, solucionando el problema en menos de 1 segundo

#### **Python**

In [55]:
%%writefile max_pairwise_product1_eficiente.py
# python3

def max_pairwise_product_eficiente(numbers):
    n = len(numbers)
    max_product = 0
    mayor = 0
    smayor=0
    for first in range(n):
      if numbers[first]>mayor:
        smayor = mayor
        mayor = numbers[first]
      elif numbers[first]>smayor:
        smayor = numbers[first]
      #print("Mayor: ",mayor,"  -  Segundo: ",smayor,"\n")
    max_product = mayor*smayor
    return max_product

if __name__ == '__main__':
    input_n = int(input())
    input_numbers = [int(x) for x in input().split()]
    print(max_pairwise_product_eficiente(input_numbers))


Writing max_pairwise_product1_eficiente.py


##### Test 1: Small numbers

In [56]:
!time python max_pairwise_product1_eficiente.py<input1.txt

90

real	0m0.068s
user	0m0.044s
sys	0m0.019s


##### Test 2: largest input numbers 200000 200000 in the input

In [57]:
!time python max_pairwise_product1_eficiente.py<input2.txt

40000000000

real	0m0.063s
user	0m0.042s
sys	0m0.018s


##### Test 3  stress tests 20  times  random n in [2, 20]

In [58]:
!time python max_pairwise_product1_super_efficient.py<input3.txt

python3: can't open file 'max_pairwise_product1_super_efficient.py': [Errno 2] No such file or directory

real	0m0.062s
user	0m0.035s
sys	0m0.017s


##### Test 4  time test  measure the execution time running with the the largest worst case file (200000 increasing numbers, 200000 and 200000 the end) 

In [59]:
!time python max_pairwise_product1_super_efficient.py<input4.txt

python3: can't open file 'max_pairwise_product1_super_efficient.py': [Errno 2] No such file or directory

real	0m0.071s
user	0m0.046s
sys	0m0.015s


Por lo que se puede evidenciar que es un algoritmo bastante optimizado

## **Final super efficient solution (without storing the numbers , without calling a function)**

#### **C++**

Se va a modificar el algoritmo de c++ para que sea más eficiente, esto se va a hacer sin necesidad de almacenar los datos.

In [60]:
%%writefile max_pairwise_product_super_eficiente.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <list>


int main() {
    int n;
    long long max_product = 0LL;
    long long mayor = 0LL;
    long long smayor = 0LL;
    long long temp = 0LL;
    std::cin >> n;
    for (int i = 0; i < n; ++i) {
        std::cin >> temp;
        if(temp>mayor){
          smayor = mayor;
          mayor = temp;
        }else if(temp>smayor){
          smayor = temp;
        }
    }
    max_product = (long long)mayor*smayor;
    std::cout<<max_product<<"\n";
    return 0;
}



Writing max_pairwise_product_super_eficiente.cpp


In [61]:
!g++ max_pairwise_product_super_eficiente.cpp -o max_pairwise_product_super_eficiente

##### Test 1: Small numbers

In [62]:
!./max_pairwise_product < input1.txt

90


##### Test 2: largest input numbers 200000 200000 in the input

In [63]:
!./max_pairwise_product < input2.txt

1345294336


##### Test 3  stress tests 20  times  random n in [2, 20] 

In [64]:
!time ./max_pairwise_product_long <input3.txt

35802189619

real	0m0.007s
user	0m0.002s
sys	0m0.001s


##### Test 4 time test measure the execution time running with the the largest worst case file (200000 increasing numbers, 200000 and 200000 the end)

In [65]:
! time ./max_pairwise_product_super_eficiente<input4.txt

40000200000

real	0m0.077s
user	0m0.074s
sys	0m0.002s


#### **Python**

Se trata de la misma solución eficiente, solamente que en este caso no se almacenan los valores en una lista, sino que se comparan directamente al ser recibidos.

In [71]:
%%writefile max_pairwise_product1_super_efficient.py
# python3

if __name__ == '__main__':
    max_product = 0
    mayor = 0
    smayor = 0
    temp = 0
    input_n = int(input())
    for x in input().split():
      temp = int(x)
      if temp>mayor:
        smayor = mayor
        mayor = temp
      elif temp>smayor:
        smayor = temp
    max_product = mayor*smayor
    print(max_product)

Overwriting max_pairwise_product1_super_efficient.py


##### Test 1: Small numbers

In [72]:
!time python max_pairwise_product1_super_efficient.py<input1.txt

90

real	0m0.057s
user	0m0.038s
sys	0m0.014s


##### Test 2: largest input numbers 200000 200000 in the input

In [73]:
!time python max_pairwise_product1_super_efficient.py<input2.txt

40000000000

real	0m0.062s
user	0m0.040s
sys	0m0.019s


##### Test 3  stress tests 20  times  random n in [2, 20]

In [74]:
!time python max_pairwise_product1_super_efficient.py<input3.txt

35802189619

real	0m0.072s
user	0m0.045s
sys	0m0.020s


##### Test 4  time test  measure the execution time running with the the largest worst case file (200000 increasing numbers, 200000 and 200000 the end) 

In [75]:
!time python max_pairwise_product1_super_efficient.py<input4.txt

40000200000

real	0m0.194s
user	0m0.162s
sys	0m0.022s


#### **Count**

Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
# Poner codigo eficiente con conteo de instrucciones

Calculate estimated time by instruction and number of instructions by second

In [ ]:
# Completar

##### Test 1: Small numbers

In [ ]:
!time python max_pairwise_product1_super_efficient.py<input1.txt

90

real	0m0.058s
user	0m0.041s
sys	0m0.016s


##### Test 2: largest input numbers 200000 200000 in the input

In [ ]:
!time python max_pairwise_product1_super_efficient.py<input2.txt

40000000000

real	0m0.056s
user	0m0.039s
sys	0m0.018s


##### Test 3  stress tests 20  times  random n in [2, 20]

In [ ]:
!time python max_pairwise_product1_super_efficient.py<input3.txt

38776289740

real	0m0.062s
user	0m0.036s
sys	0m0.025s


Parece que funciona bastante bien, se hará la prueba con 200.000 datos, con por lo menos dos 200.000 en sus entradas:

##### Test 4  time test  measure the execution time running with the the largest worst case file (200000 increasing numbers, 200000 and 200000 the end) 

In [ ]:
!time python max_pairwise_product1_super_efficient.py<input4.txt

40000200000

real	0m0.151s
user	0m0.120s
sys	0m0.029s


Se aprecia una disminución considerable en el tiempo de ejecución, aproximadamente de 0.014s, que, aunque no sea muy distinguible, es un tiempo considerable, el código super optimizado es 7,48% más rápido que el código optimizado.

# **Captura de pantalla coursera**

<img src ="https://raw.githubusercontent.com/drriosg/AlgorithmsUN2021I/master/Lab6/drriosgAlgorithmicToolboxWeek1.jpg">